[NodeGen] Support enum members on nodes

SplitInfinity opened the pull request:
**Description**
This commit adds support for enum members on nodes. An enum type
can be defined within `NodeGen` itself, or defined externally.
In the latter case, it must be declared in `NodeGen` so
that it can be forward declared correctly when defining `Nodes` that
have members of that type.

The API looks something like this:

```
auto e = BB.newEnum("glow::greatBackend::tremendousEnum");
e.addValue("FIRST").addValue("SECOND");
BB.newNode(...).addMember(e, "EnumMember");
```

**Testing**
Tried out new features with a new autogenerated node class with four
different enum members, two of one type and two of another. The two enum
types were declared in different namespaces:

```
  auto &e = BB.newEnum("newbackend::namespace1::enum1");
  e.addEnumCase("FIRST", "first comment");
  e.addEnumCase("SECOND", "second comment");
  e.setDocstring("This is a test docstring. Lorem ipsum doloret something "
                 "something.");
  auto eType = e.getMemberTypeInfo();

  auto &f = BB.newEnum("newbackend::namespace1::namespace2::enum2");
  f.addEnumCaseWithValue("FIRST", 5);
  f.addEnumCaseWithValue("SECOND", 10);
  f.addEnumCaseWithValue("THIRD", 1);
  f.addEnumCase("FOURTH");
  f.setDocstring("This is a test docstring. Lorem ipsum doloret something "
                 "something.");
  auto fType = f.getMemberTypeInfo();

  BB.newNode("EnumMemberTest")
      .addInput("Input")
      .addResultFromCtorArg()
      .addMember(&eType, "EnumMember1")
      .addMember(&eType, "EnumMember2")
      .addMember(MemberType::Float, "someFloat")
      .addMember(&fType, "EnumMember3")
      .addMember(&fType, "EnumMember4");
```

Everything compiled and the generated code looked correct.

**Fixes**
Fixes issue #2053.

SplitInfinity commented on tools/ClassGen/NodeBuilder.h:
```diff
@@ -229,6 +308,18 @@ class Builder {
   }
 
+  /// Declare a new enum and generate code for it.
+  EnumBuilder *newEnum(const std::string &name) {
```
I made this return a pointer because my understanding of r-values and move semantics is not strong enough for me to express this in those terms. If I make this return an `EnumBuilder`, the destructor gets called twice (once for for destroying the copy made by this function, another time for destroying the copy that resides in the caller's scope) which causes the enum definition to be emitted twice.

Some help expressing this without pointers would be awesome.

jackm321 replied:
I'm not sure I understand your comment exactly but if you return an `EnumBuilder` from this function directly it should be no problem and then use an rvalue reference for addMember like `NodeBuilder &addMember(EnumBuilder &&eb, const std::string &name)` (and also move into `enumMembers_ ` inside of addMember) it should fix the problem.

jackm321 replied:
This will only allow an enum to be used in one node though so it may not be what you want.

SplitInfinity replied:
Yeah, one of the goals of this change is to allow multiple nodes to have a member of the same type of enum, so this solution is not going to work.

bertmaher replied:
I kinda feel like this is a symptom of an overly-clever use of the destructor to emit code... I wouldn't mind an explicit `emit()` method.

SplitInfinity commented on tools/ClassGen/NodeBuilder.h:
```diff
@@ -229,6 +308,18 @@ class Builder {
+  }
+
+  /// Declare a new backend specific enum. This will NOT generate a definition.
+  EnumBuilder *newBackendSpecificEnum(const std::string &name) {
```
Same here.

opti-mix commented:
@SplitInfinity Thanks for working on this feature! It is going to be very useful.

I've got a couple of requests for you.

The easy ones:
- Could you allow for setting doc-strings for enumerators as well?
  May be using an API like `addCase(std::string enumeratorName, std::string enumeratorDoc = "")`, so that the doc-string is optional.

- Could you allow for specifying a numeric value for the enumerators? 
  May be using an API like `addCaseWithValue(std::string enumeratorName, int enumeratorValue)` or something like this?

The more complex ones:
- Regarding using the enums declared externally: you say that they need to be forward declared. Does it need to be done using NodeGen APIs? Or do you need to write (or include) a usual C++ forward declaration? Do you need to/Can you include any backend-specific header files defining the actual type? Can your approach be used only for enums or could it be extended to support any other externally defined types? It could be very useful to be able to use externally defined types as types of the members.

- We may need to add the same features to the InstrGen, because e.g. when we lower a node into a backend-specific instruction, we most likely want to propagate the members of custom types into this instruction.

opti-mix commented on tools/ClassGen/NodeBuilder.h:
```diff
@@ -45,6 +110,8 @@ class NodeBuilder {
   /// A list of node members. Format: (type, name).
   std::vector<std::pair<MemberType, std::string>> members_;
+  /// A list of external enum node members. Format: (EnumBuilder*, name).
+  std::vector<std::pair<const EnumBuilder *, std::string>> enumMembers_;
```
You add enum members to this separate list, different from `members_`. Therefore, the order of members in the generated class will be different from the order of members as it is provided in the NodeGen. May be such members should be added to the `members_`? E.g. `MemberType` would become something more complex than a simple enum describing a predefined set of types and  `EnumBuilder` should inherit from `MemberType`.

nadavrot commented:
Nice!

SplitInfinity commented:
Changes since last push:
- added support for enumeration documentation
- added support for custom enumeration values

SplitInfinity commented on tools/ClassGen/NodeBuilder.cpp:
```diff
@@ -16,6 +16,125 @@
+}
+
+EnumBuilder &EnumBuilder::addEnumeration(const std::string &name,
+                                         const std::string &doc) {
```
It does have a default value and it is the empty string. It is specified in the declaration of the function in `NodeBuilder.h`.

SplitInfinity commented on tools/ClassGen/NodeBuilder.cpp:
```diff
@@ -16,6 +16,125 @@
+
+EnumBuilder &EnumBuilder::addEnumerationWithValue(const std::string &name,
+                                                  const int value,
+                                                  const std::string &doc) {
```
It does have a default value and it is the empty string. It is specified in the declaration of the function in `NodeBuilder.h`.

opti-mix commented on tools/ClassGen/NodeBuilder.h:
```diff
@@ -30,6 +30,73 @@
+  }
+
+  /// Add a new enumeration to the enumeration.
+  EnumBuilder &addEnumeration(const std::string &name,
```
I'd call it `addCase` or `addEnumerator`. Enumeration is usually used to refer to the whole enum.

SplitInfinity replied:
Okay, I've fixed this locally and it will be included with my next push of major revisions.

SplitInfinity commented:
Changes since last push:
- renamed `addEnumeration...` to `addEnumCase...`
- modified `MemberType` so that the order of constructor arguments matches the declaration order in `NodeGen`. I'd like some feedback on the way in which I implemented this before I modify `InstrGen` to comply with these changes.

opti-mix commented on tools/ClassGen/MemberType.cpp:
```diff
@@ -0,0 +1,38 @@
+                         "std::vector<size_t>", "std::vector<size_t>")
+DEFINE_BASIC_MEMBER_TYPE(VectorNodeValue, "NodeValueArrayRef",
+                         "std::vector<NodeHandle>", "std::vector<NodeValue>")
+
```
Looks like you do not win a lot by using macro definitions here. You could just as well define these types directly.

bertmaher commented on tools/ClassGen/NodeBuilder.h:
```diff
@@ -30,6 +30,73 @@
+  /// The containing namespaces for the enum.
+  std::vector<std::string> namespaces_;
+  /// The enum case names, values, and their corresponding docstrings.
+  std::vector<std::tuple<std::string, int, std::string>> enumCases_;
```
I'd kinda prefer a struct with {name, value, docstring} over std::tuple.

SplitInfinity replied:
Fixed in latest push.

jackm321 commented on tools/ClassGen/NodeBuilder.cpp:
```diff
@@ -16,6 +16,169 @@
+  enumCases_.emplace_back(std::make_tuple(name, value, doc));
+  // Maintain the invariant that unusedValue_ is always a value that is NOT
+  // used by the enum.
+  unusedValue_ = std::max(unusedValue_, value + 1);
```
This may not matter much but this doesn't prevent multiple cases with the same value if someone makes n cases with `addEnumCase()` then calls `addEnumCaseWithValue()` with value < n expecting that this will automatically avoid conflicts. Is it worth it to keep a set of values and assert that new cases aren't in that?

jackm321 replied:
Even just `assert(value >= unusedValue_)` would work then just make `unusedValue_ = value + 1`

SplitInfinity replied:
I sort of fixed this by scanning for and fixing duplicate values in `addEnumCaseWithValue` if a user-defined value conflicts with a value assigned by the tool. It's a linear time method instead of a constant time method now, though.

SplitInfinity commented:
Changes since last push:
- replace `std::tuple` used to represent enum cases with an `EnumCase` struct
- handled case in which `addEnumCaseWithValue` is called with a value that has already been assigned by `EnumBuilder` during a previous call to `addEnumCase`. This way, mixing `addEnumCase` and `addEnumCaseWithValue` will "just work", but the tool will silently generate legal code if `addEnumCaseWithValue` is called twice with the same value and different names.

SplitInfinity commented:
@opti-mix 

Seems like a good candidate for LIT.

nadavrot commented:
Nice!
